src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs

```diff
@@ -159,45 +172,46 @@ public override bool GetMemberAndWriteJsonExtensionData(object obj, ref WriteSta
         public override bool ReadJsonAndSetMember(object obj, ref ReadStack state, ref Utf8JsonReader reader)
         {
             bool success;
+
             bool isNullToken = reader.TokenType == JsonTokenType.Null;
             if (isNullToken && !Converter.HandleNull && !state.IsContinuation)
             {
-                if (!IgnoreNullValues)
+                if (!Converter.CanBeNull)
                 {
-                    if (!Converter.CanBeNull)
-                    {
-                        ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Converter.TypeToConvert);
-                    }
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Converter.TypeToConvert);
+                }
+
+                Debug.Assert(s_defaultValue == null);
 
+                if (!IgnoreDefaultValuesOnRead)
+                {
                     T value = default;
                     Set!(obj, value!);
                 }
 
                 success = true;
             }
-            else
+            else if (Converter.CanUseDirectReadOrWrite)
             {
-                // Get the value from the converter and set the property.
-                if (Converter.CanUseDirectReadOrWrite)
+                if (!(isNullToken && IgnoreDefaultValuesOnRead && Converter.CanBeNull))
                 {
                     // Optimize for internal converters by avoiding the extra call to TryRead.
                     T fastvalue = Converter.Read(ref reader, RuntimePropertyType!, Options);
-                    if (!IgnoreNullValues || (!isNullToken && fastvalue != null))
-                    {
-                        Set!(obj, fastvalue!);
-                    }
-
-                    return true;
+                    Set!(obj, fastvalue!);
```

Comment:
The semantics here changed when using IgnoreNullValues. Previously, the setter wasn't called when the value is null. 

If we keep this, we should document it as a breaking change.

Reply:
I think we should keep this and document the change. The only way for `null` to be set here is if the token type was _not_ `JsonTokenType.Null` and the converter returned null for some reason. I've added a test to express this behavior.

AFAIK, `IgnoreNullValues` means if the token type for the property is `JsonTokenType.Null` when deserializing, _and_ the type to deserialize can be `null`, then the serializer should skip it. We still honor this contract.